packages/autorest.python/autorest/cadlflags/__init__.py

```diff
@@ -0,0 +1,128 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+from typing import Any, Dict, List
+from .. import YamlUpdatePlugin
+from .._utils import parse_args
+
+_LOGGER = logging.getLogger(__name__)
+
+OAUTH_TYPE = "OAuth2"
+KEY_TYPE = "Key"
+
+
+def get_azure_key_credential(key: str) -> Dict[str, Any]:
+    return {
+        "type": KEY_TYPE,
+        "policy": {"type": "AzureKeyCredentialPolicy", "key": key},
+    }
+
+
+class CadlFlags(YamlUpdatePlugin):  # pylint: disable=abstract-method
+    """A plugin to apply flags from backdoor python.json into cadl yaml file"""
+
+    def update_yaml(self, yaml_data: Dict[str, Any]) -> None:
+        """Convert in place the YAML str."""
+        if self.options.get("add-credential"):
+            self.update_credential(yaml_data)
+        if self.options.get("namespace"):
+            yaml_data["client"]["namespace"] = self.options["namespace"]
+        if self.options.get("title"):
+            yaml_data["client"]["name"] = self.options["title"]
+
+    def get_credential_scopes_from_flags(self, auth_policy: str) -> List[str]:
+        if self.options.get("azure-arm"):
+            return ["https://management.azure.com/.default"]
+        credential_scopes_temp = self.options.get("credential-scopes")
+        credential_scopes = (
+            credential_scopes_temp.split(",") if credential_scopes_temp else None
+        )
+        if self.options.get("credential-scopes", False) and not credential_scopes:
+            raise ValueError(
+                "--credential-scopes takes a list of scopes in comma separated format. "
+                "For example: --credential-scopes=https://cognitiveservices.azure.com/.default"
+            )
+        if not credential_scopes:
+            _LOGGER.warning(
+                "You have default credential policy %s "
+                "but not the --credential-scopes flag set while generating non-management plane code. "
+                "This is not recommend because it forces the customer to pass credential scopes "
+                "through kwargs if they want to authenticate.",
+                auth_policy,
+            )
+            credential_scopes = []
+        return credential_scopes
+
+    def get_token_credential(self, credential_scopes: List[str]) -> Dict[str, Any]:
+        return {
+            "type": OAUTH_TYPE,
+            "policy": {
+                "type": "ARMChallengeAuthenticationPolicy"
+                if self.options.get("azure-arm")
+                else "BearerTokenCredentialPolicy",
+                "credentialScopes": credential_scopes,
+            },
+        }
+
+    def update_credential_from_flags(self) -> Dict[str, Any]:
```

Comment:
Will it cause some duplicated code in [m4reformatter](https://github.com/Azure/autorest.python/blob/88ca41a646740bc66fdfc09940244e3b2ecd9f8b/packages/autorest.python/autorest/m4reformatter/__init__.py#L970-L981)?

Reply:
it's semi-duplicated, but it's still kind of different. This plugin will also be removed in the hopefully near future, so not worth it to refactor all of the code to avoid duplication imo